onnx/backend/test/case/node/scalardotproductattention.py

```diff
@@ -0,0 +1,1034 @@
+    # NewShapeV (value) has shape (batch_size, kv_num_heads, kv_sequence_length, v_head_size)
+    batch_size = Q.shape[0]
+    if len(Q.shape) == 3:
+        hidden_size_q = Q.shape[2]
+        hidden_size_k = K.shape[2]
+        hidden_size_v = V.shape[2]
+        assert q_num_heads is not None and kv_num_heads is not None
+
+        head_size_q = int(hidden_size_q / q_num_heads)
+        new_shape_q = [batch_size, q_num_heads, Q.shape[1], head_size_q]
+        Q = np.reshape(Q, new_shape_q)
+
+        head_size_k = int(hidden_size_k / kv_num_heads)
+        new_shape_k = [batch_size, kv_num_heads, K.shape[1], head_size_k]
+        K = np.reshape(K, new_shape_k)
+
+        head_size_v = int(hidden_size_v / kv_num_heads)
+        new_shape_v = [batch_size, kv_num_heads, V.shape[1], head_size_v]
+        V = np.reshape(V, new_shape_v)
+    assert len(Q.shape) == 4 and len(K.shape) == 4 and len(V.shape) == 4
+
+    # Calculate Scaling Factor if not provided
+    if scale is None:
+        q_head_size = Q.shape[3]
+        scale = 1 / np.sqrt(q_head_size)
+
+    # Update key and value cache
+    if past_key is not None:
+        present_key = np.concatenate((past_key, K), axis=2)
+    else:
+        present_key = K
+    if past_value is not None:
+        present_value = np.concatenate((past_value, V), axis=2)
+    else:
+        present_value = V
+    K = present_key
+    V = present_value
+
+    # Create attn_bias
+    q_sequence_length = Q.shape[2]
+    kv_sequence_length = K.shape[2]
+    attn_bias = np.zeros((q_sequence_length, kv_sequence_length), dtype=Q.dtype)
+    # First case: If is_causal is provided
+    # If set to true, the attention masking is a lower triangular matrix when the mask
+    # is a square matrix. The attention masking has the form of the upper left causal
+    # bias due to the alignment when the mask is a non-square matrix.
+    if is_causal is True:
+        assert attn_mask is None
+        temp_mask = np.ones((q_sequence_length, kv_sequence_length), dtype=np.bool)
+        temp_mask = np.tril(temp_mask, k=0)
+        temp_mask = np.logical_not(temp_mask)
+        attn_bias_ma = np.ma.array(attn_bias, mask=temp_mask)
+        attn_bias = attn_bias_ma.filled(fill_value=float("-inf"))
+    if attn_mask is not None:
+        assert is_causal is False
+        if attn_mask.dtype == np.bool:
+            attn_mask = np.logical_not(attn_mask)
+            attn_bias_ma = np.ma.array(attn_bias, mask=attn_mask)
+            attn_bias = attn_bias_ma.filled(fill_value=float("-inf"))
+        else:
+            attn_bias += attn_mask
+
+    # Group Query Attention is applied if the following are satisfied
+    # 1) q_num_heads != kv_num_heads
+    # 2) q_num_heads % kv_num_heads == 0
+    # 3) kv_num_heads == k_num_heads == v_num_heads
+    if q_num_heads is None:
+        q_num_heads = Q.shape[1]
+    if kv_num_heads is None:
+        k_num_heads = K.shape[1]
+        v_num_heads = K.shape[1]
```

Comment:
Is there any reason to separate the kv_num_heads into two values instead of just using kv_num_heads?

instead of 

```
 if kv_num_heads is None:
     k_num_heads = K.shape[1]
      v_num_heads = K.shape[1]
else:
      k_num_heads = kv_num_heads
      v_num_heads = kv_num_heads
```

```
 if kv_num_heads is None:
     kv_num_heads = K.shape[1]
```

then just use `kv_num_heads` for all the calculations bellow.

With this change the else is removed. and the line 108 can be deleted because by definition kv_num_heads is equal to itself.